File: src/api/diet.ts
import { getBearerAuthHeaders } from './axiosConfig.js'
import { apiFetch } from './client.ts'

export type MealSlot = 'breakfast' | 'lunch' | 'dinner' | 'snack_1' | 'snack_2'

export type MealLogEntry = {
  meal_slot: MealSlot
  description: string
  calories: number | null
  protein_g?: number | null
  carbs_g?: number | null
  fat_g?: number | null
}

export type TodayMealLog = {
  date: string
  entries: MealLogEntry[]
  notes: string
}

export type PlannedMeal = {
  title: string
  prep_minutes: number
  ingredients: string[]
  notes: string
}

export type DietPlan = {
  id: string
  name: string
  calorie_goal: number
  goal_type: 'weight_loss' | 'muscle_gain' | 'maintenance'
  meals: {
    breakfast: PlannedMeal
    lunch: PlannedMeal
    dinner: PlannedMeal
    snack_1: PlannedMeal
    snack_2: PlannedMeal
  }
  macros: {
    protein: number
    carbs: number
    fat: number
  }
}

export type DietLogResult = {
  ok: boolean
  date: string
  entry_count: number
  message?: string | null
}

/** GET /api/diet/{calorie_goal} — plans near this calorie target */
export async function fetchDietPlansByCalorieGoal(calorieGoal: number): Promise<DietPlan[]> {
  const h = await getBearerAuthHeaders()
  return apiFetch<DietPlan[]>(`/api/diet/${calorieGoal}`, { headers: h })
}

export async function fetchTodayMealLog(): Promise<TodayMealLog> {
  const h = await getBearerAuthHeaders()
  return apiFetch<TodayMealLog>('/api/diet/log/today', { headers: h })
}

export async function saveMealLog(entries: MealLogEntry[], notes = ''): Promise<DietLogResult> {
  const h = await getBearerAuthHeaders()
  return apiFetch<DietLogResult>('/api/diet/log', {
    method: 'POST',
    headers: h,
    body: JSON.stringify({ entries, notes }),
  })
}

File: backend/app/schemas/diet.py
"""API schemas for diet endpoints."""

from __future__ import annotations

from typing import List, Literal

from pydantic import BaseModel, Field

from app.models.diet_plan import DietGoalType, DietMacros, DietMealsStructured


class DietPlanOut(BaseModel):
    """Full diet plan for API responses."""

    id: str
    name: str
    calorie_goal: int
    goal_type: DietGoalType
    meals: DietMealsStructured
    macros: DietMacros


class MealLogEntryIn(BaseModel):
    meal_slot: Literal["breakfast", "lunch", "dinner", "snack_1", "snack_2"]
    description: str = Field(..., min_length=1, description="What was eaten")
    calories: int | None = Field(default=None, ge=0, description="Estimated kcal for this entry")
    protein_g: float | None = Field(default=None, ge=0, description="Protein (g) for this entry")
    carbs_g: float | None = Field(default=None, ge=0, description="Carbs (g) for this entry")
    fat_g: float | None = Field(default=None, ge=0, description="Fat (g) for this entry")


class MealLogEntryOut(BaseModel):
    meal_slot: str
    description: str
    calories: int | None = None
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None


class TodayMealLogOut(BaseModel):
    date: str
    entries: list[MealLogEntryOut]
    notes: str = ""


class DietLogIn(BaseModel):
    """Replace or create today's meal log (UTC day)."""

    entries: list[MealLogEntryIn] = Field(
        default_factory=list,
        description="Meals eaten today; can be empty to clear notes-only logs",
    )
    notes: str = Field(default="", description="Optional day notes")


class DietLogOut(BaseModel):
    ok: bool = True
    date: str = Field(..., description="UTC date ISO (YYYY-MM-DD)")
    entry_count: int
    message: str | None = None
